plugins/statistics.koplugin/main.lua

```diff
@@ -73,58 +86,429 @@ function ReaderStatistics:init()
     if self:isDocless() or not self.is_enabled then
         return
     end
     -- first execution
     if not self.data then
         self.data = { performance_in_pages= {} }
-        self:inplaceMigration();  -- first time merge data
     end
-
     local book_properties = self:getBookProperties()
     self.data.title = book_properties.title
+    if self.data.title == nil or self.data.title == "" then
+        self.data.title = self.document.file:match("^.+/(.+)$")
+    end
     self.data.authors = book_properties.authors
     self.data.language = book_properties.language
     self.data.series = book_properties.series
 
     self.data.pages = self.view.document:getPageCount()
-    return
+    self.data.md5 = self:md5(self.document.file)
+    self.curr_total_time = 0
+    self.curr_total_pages = 0
+    self.id_curr_book = self:getIdBookDB()
+    self.total_read_pages, self.total_read_time = self:getPageTimeTotalStats(self.id_curr_book)
+    if self.total_read_pages > 0 then
+        local avg_time = math.ceil(self.total_read_time / self.total_read_pages)
+        self.avg_time = avg_time
+        self.ui:handleEvent(Event:new("UpdateStats", avg_time))
+        self.view.footer:updateFooter()
+    end
 end
 
-local function generateReadBooksTable(title, dates)
-    local result = {}
-    for k, v in tableutil.spairs(dates, function(t, a, b) return t[b].date < t[a].date end) do
-        table.insert(result, {
-            k,
-            T(_("Pages (%1) Time: %2"), v.count, util.secondsToClock(v.read, false))
-        })
+function ReaderStatistics:checkInitDatabase()
+    local conn = SQ3.open(db_location)
+    if self.convert_to_db then      -- if conversion to sqlite was doing earlier
+        if not conn:exec("pragma table_info('book');") then
+            UIManager:show(ConfirmBox:new{
+                text = T(_("Cannot open database in %1.\nProbably the database has been deleted or moved.\n " ..
+                    "Do you want to create empty database?"), db_location),
+                cancel_text = _("Close"),
+                cancel_callback = function()
+                    return
+                end,
+                ok_text = _("Create"),
+                ok_callback = function()
+                    local conn_new = SQ3.open(db_location)
+                    self:createDB(conn_new)
+                    conn_new:close()
+                    UIManager:show(InfoMessage:new{text =_("A new empty database has been created."), timeout = 3 })
+                    self:initData()
+                end,
+            })
+        end
+    else  -- first time convertion to sqlite database
+        self.convert_to_db = true
+        if not conn:exec("pragma table_info('book');") then
+            if #ReadHistory.hist > 0 then
+                local info = InfoMessage:new{
+                    text =_("New version of statistics plugin is detected.\n" ..
+                        "You need to convert statistics data to database.\n"..
```

Comment:
"You need" implies user action. This should either be changed to a confirm dialog (so that users can opt out of the process) or it should just say something like "Statistics data needs to be converted into the new database format."